content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+  AssumeSingleThreaded::new(LeakingAllocator::new());
+
+// "Proper" allocator with added thread-safety
+#[global_allocator]
+static ALLOCATOR: LockedAllocator<FreeListAllocator> = 
+  LockedAllocator::new(FreeListAllocator::new());
+
+// ...
+```
+
+I don’t have any first-hand experience with `lol_alloc` outside of writing this article, but it looks good!
+
+## wasm-bindgen
+
+Now that we've done pretty much everything the hard way, we have earned a look at the easy way of writing Rust for WebAssembly, which is using [wasm-bindgen].
+
+They key feature of wasm-bindgen is the `#[wasm_bindgen]` macro that you put on every function that you want to be exported. This macro adds the same compiler directives we added manually earlier in this article, but it does something way more useful in addition to that:
+
+For example, if we add the macro to our `add` function from above, it emits another function called `__wbindgen_describe_add` and exports it. This function returns the a description of the function's signature in a [numeric format][wasm-bindgen descriptor]. The descriptor of our `add` function looks like this:
+
+```
+Function(
+    Function {
+        arguments: [
+            U32,
+            U32,
+        ],
+        shim_idx: 0,
+        ret: U32,
+        inner_ret: Some(
+            U32,
+        ),
+    },
+)
+```
+
+This is quite the simple function, but descriptor data structure is capable of representing quite complex function signatures. 
+
+> **Expand:** If you want to see what code the `#[wasm_bindgen]` macro emits, use rust-analyzer's "Expand Macro recursively" functionality. You can run it via VS Code through the Command Palette.
+
+wasm-bindgen is not just a crate, but also comes with a CLI, that you run as a post-processing step on our Wasm binary. The CLI extracts those descriptors and uses the information to generate tailor-made JavaScript bindings (often also called “glue code”), and then removes all these descriptor functions from the binary. The JavaScript encodes all the knowledge about higher-level types I talked about earlier, allowing you to seamlessly pass types like strings, `ArrayBuffer` or even closures.
+\
+If you want to write Rust for WebAssembly, wasm-bindgen should be your first choice. wasm-bindgen doesn’t work with `#![no_std]`, but in practice that is rarely a problem.
+
+### Custom Section
+
+Fun fact: We can use Rust to add our own custom sections to a WebAssembly module. If we declare an array of bytes (not a slice!), we can add a `#[link_section = ...]` attribute to pack those bytes into its own section.
+
+```rust
+const _: () = {
+    #[link_section = "surmsection"]
+    static SECTION_CONTENT: [u8; 11] = *b"hello world";
+};
+```
+
+And we can extract this data using the [`WebAssembly.Module.customSection()` API][customsection] or using `wasm-objdump`:
+
+```
+$ wasm-objdump -s -j surmsection target/wasm32-unknown-unknown/release/my_project.wasm
```

Comment:
This could be a good example for `wasm-dis` btw which outputs custom sections in a slightly more readable form:

```
> wasm-dis temp.wasm
(module
…
 ;; custom section "surmsection", size 11, contents: "hello world"
)
```

Reply:
But only if it is a printable string. If it's binary, it doesn't show anything.

Reply:
Yeah I just like that it doesn't join section name and data together. Maybe a good feature request for `wasm-objdump`, there is no reason why `-s` or `-x` wouldn't print name & data separately.

Reply:
Yeah I just like that it doesn't join section name and data together. Maybe a good feature request for `wasm-objdump`, there is no reason why `-s` or `-x` wouldn't print name & data separately.